### ISSUE TITLE
Patch mainTemplate.gradle is incompatible with Export Project

### ISSUE THREAD

**Issue by stopiccot** [firebase/quickstart-unity#393]
## Please fill in the following fields:
Unity editor version: 2018.4.0
Firebase Unity SDK version:  6.1.1
Platform you are using the Unity editor on (Mac, Windows, or Linux):  Mac
Platform you are targeting (iOS, Android, and/or desktop): Android

## Please describe the issue here:
- intialized empty unity project
- copied mainTemplate.gradle from `2018.4.0f1/PlaybackEngines/AndroidPlayer/Tools/GradleTemplates/`
- added FirebaseAnalytics package from 6.1.1 sdk
- exported android studio project

Upon opening project in Android Studio 3.2.1 I get following error during gradle sync:
```
Failed to resolve: com.google.firebase:firebase-analytics-unity:6.1.1
Show in File
Show in Project Structure dialog


Failed to resolve: com.google.firebase:firebase-app-unity:6.1.1
Show in File
Show in Project Structure dialog
```

Seems like `url (unityProjectPath + "/Assets/Firebase/m2repository") ` hack is not working.
To be honest I do not understand why this local m2 repo is needed at all. Why `-unity` firebase aars can't be simply fetched from maven like other firebase aars do.

**Comment by patm1987** [comment]
Hi @stopiccot , thanks for taking the time to file a bug!

I can't quite seem to replicate your error in either Unity 2018.4.1 or Unity 2019.1.6. My only variation is that I tend to check the "use custom gradle file" checkbox under publishing settings to auto-copy the mainTemplate.gradle project for me.

Do you have roughly a section in your build.gradle that looks like this:
```
// Android Resolver Dependencies Start
    implementation 'com.google.auto.value:auto-value-annotations:1.6.3' // Assets/Firebase/Editor/AppDependencies.xml:22
    implementation 'com.google.firebase:firebase-analytics:16.5.0' // Assets/Firebase/Editor/AppDependencies.xml:15
    implementation 'com.google.firebase:firebase-analytics-unity:6.1.1' // Assets/Firebase/Editor/AnalyticsDependencies.xml:20
    implementation 'com.google.firebase:firebase-app-unity:6.1.1' // Assets/Firebase/Editor/AppDependencies.xml:20
    implementation 'com.google.firebase:firebase-common:17.1.0' // Assets/Firebase/Editor/AppDependencies.xml:13
    implementation 'com.google.firebase:firebase-iid:[18.0.0]' // Assets/Firebase/Editor/AnalyticsDependencies.xml:15
// Android Resolver Dependencies End
```

And can you force resolve dependencies from the play services resolver menu:
![image](https://user-images.githubusercontent.com/1111611/59811886-b637d800-92c0-11e9-98c2-ee70f0da742f.png)

If none of that works, can you also attach your editor log if you remain stuck?

--Patrick

**Comment by stopiccot** [comment]
Yes, I have mentioned section in my mainTemplate.gradle after resolution completes. Here is my dependencies section:
```
dependencies {
    implementation fileTree(dir: 'libs', include: ['*.jar'])
// Android Resolver Dependencies Start
    implementation 'com.google.auto.value:auto-value-annotations:1.6.3' // Assets/Firebase/Editor/AppDependencies.xml:22
    implementation 'com.google.firebase:firebase-analytics:16.5.0' // Assets/Firebase/Editor/AppDependencies.xml:15
    implementation 'com.google.firebase:firebase-analytics-unity:6.1.1' // Assets/Firebase/Editor/AnalyticsDependencies.xml:20
    implementation 'com.google.firebase:firebase-app-unity:6.1.1' // Assets/Firebase/Editor/AppDependencies.xml:20
    implementation 'com.google.firebase:firebase-common:17.1.0' // Assets/Firebase/Editor/AppDependencies.xml:13
    implementation 'com.google.firebase:firebase-iid:[18.0.0]' // Assets/Firebase/Editor/AnalyticsDependencies.xml:15
// Android Resolver Dependencies End
    implementation project(':Firebase')
}
```

After that I export android studio project from unity:

<img width="629" alt="Screenshot 2019-06-20 at 10 59 42" src="https://user-images.githubusercontent.com/24836/59831505-8e8d5200-934a-11e9-88af-31d19325887a.png">

After successfull export I open this project using Android Studio 3.2.1.
But Android Studio fails to open the project:

<img width="847" alt="Screenshot 2019-06-20 at 11 04 25" src="https://user-images.githubusercontent.com/24836/59831817-34d95780-934b-11e9-85dd-4fe3056e4b9f.png">

**Comment by stopiccot** [comment]
I really don't understand why `com.google.firebase:firebase-analytics:16.5.0` is located at maven.google.com but `com.google.firebase:firebase-analytics-unity:6.1.1` is located in some local mvan repository that is not working when you choose to export project instead of building apk directly.

**Comment by patm1987** [comment]
Since I haven't been able to replicate your exact issue, would you mind attaching your editor log? On MacOS, you can find this in ~/Library/Logs/Unity/Editor.log

You mentioned the local maven repository, is it present in your build.gradle and is it pointing to a valid path?
ex: mine is:
```
// Android Resolver Repos Start
([rootProject] + (rootProject.subprojects as List)).each { project ->
    project.repositories {
        def unityProjectPath = "file:///" + file(rootProject.projectDir.path + "/../../").absolutePath
        maven {
            url "https://maven.google.com"
        }
        maven {
            url "file:////Users/martinpatrick/Documents/Unity/TestProject/Assets/Firebase/m2repository" // Assets/Firebase/Editor/AnalyticsDependencies.xml:20, Assets/Firebase/Editor/AppDependencies.xml:20
        }
        mavenLocal()
        jcenter()
        mavenCentral()
    }
}
```

**Comment by stopiccot** [comment]
Mine is:
```
// Android Resolver Repos Start
([rootProject] + (rootProject.subprojects as List)).each { project ->
    project.repositories {
        def unityProjectPath = "file:///" + file(rootProject.projectDir.path + "/../../").absolutePath
        maven {
            url "https://maven.google.com"
        }
        maven {
            url (unityProjectPath + "/Assets/Firebase/m2repository") // Assets/Firebase/Editor/AnalyticsDependencies.xml:20, Assets/Firebase/Editor/AppDependencies.xml:20
        }
        mavenLocal()
        jcenter()
        mavenCentral()
    }
}
```
Editor log is very huge. Shall I look for some particular lines in it?

And I'll repeat my question: Why local maven repo is needed at all? Why `com.google.firebase:firebase-analytics-unity:6.1.1` can not be uploaded from maven.google.com? This way all unnecessary complication would be removed.

**Comment by patm1987** [comment]
If you want to pastebin it, I'd particularly be looking for anything regarding the generation of the unityProjectPath or adding the local maven repository. You can also try grabbing the latest play services resolver from here:
https://github.com/googlesamples/unity-jar-resolver
It is interesting that it generated `unityProjectPath` on my machine then ignored it whereas yours went along and concatenated it in. You can try manually referencing the local maven repository to get unstuck (I assume it will), but that's not a satisfactory result for me and I want to make sure I have everything I need to put together a proper bug report.

Also, apologies for ignoring your local maven repo request. I've been focused on getting you past your immediate issue and just forgot to finish addressing it. There are a few major benefits of using a local maven repository, off the top of my head:
1) it makes it easier to keep the Java/C/C++/C# all in lock step. If you haven't used C#'s foreign function interface, it's actually pretty nice compared to other managed languages. Since C is the bridge though, there are a number of funky things that happen between C# and C that could create near-undetectable errors if a user managed to mismatch native and managed components.
2) The Firebase plugin works with Unity versions back into the 5 series (before the gradle build system was finalized). Firebase would probably have to wait until Unity 2019 entered LTS and 2018's LTS's support was dropped to start considering dropping "internal" build system support (internal isn't default but is an option in 2018's current LTS).

I do love the idea of moving everything into hosted maven where possible (I have a habit of hacking on the gradle file myself). The best way to build your case is to actually add your [voice over in Firebase's support page](https://firebase.google.com/support). Feel free to link to this issue as well.

**Comment by stopiccot** [comment]
Here is my [editor log](https://gist.github.com/stopiccot/b1c9035cbb157e03ba8e294b6a44680c)

**Comment by stewartmiles** [comment]
As @takaaptech discovered there was a bug in the Android Resolver when using the mainTemplate.gradle patching method that caused local repos to be ignored if the path to the Unity project contained characters that needed to be URI escaped.  See https://github.com/googlesamples/unity-jar-resolver/issues/240

This is fixed in v1.2.120 available [here](https://github.com/googlesamples/unity-jar-resolver/releases/tag/v1.2.120).

@stopiccot we're not redistributing the Firebase Unity AARs on Maven as they're Unity specific and pretty useless to any other user of Maven artifacts.  When exporting a project from Unity there are still some concerns being tracked in https://github.com/googlesamples/unity-jar-resolver/issues/239 , feel free to follow along / send a pull request, over there if you're interested.
